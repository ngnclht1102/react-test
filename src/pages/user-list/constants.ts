export const BASE_URL = 'https://jsonplaceholder.typicode.com';
export const GOOGLE_MAP_URL = 'https://maps.google.com/?q=';
