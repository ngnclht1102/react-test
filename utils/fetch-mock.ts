export function fetchMock<T>(url: string, data: Array<T>, shouldReject?: boolean): Promise<Array<T>> {
  const timeout = 20;
  return new Promise((resolve, reject) => {
    return setTimeout(() => {
      shouldReject ? reject(new Error('Error')) : resolve(data);
    }, timeout);
  });
}
