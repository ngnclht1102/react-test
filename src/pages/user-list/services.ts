import { BASE_URL } from './constants';
import { TUser } from './type';

export const Services = {
  getUsers: (): Promise<Array<TUser>> => {
    return new Promise((resolve, reject) => {
      fetch(`${BASE_URL}/users`)
        .then((response) => response.json())
        .then((json) => resolve(json))
        .catch((error) => reject(error));
    });
  },
};
