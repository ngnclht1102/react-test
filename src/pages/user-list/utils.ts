import { useCallback, useEffect, useState } from 'react';
import { GOOGLE_MAP_URL } from './constants';

export function useQuery<T>(promise: () => Promise<Array<T>>) {
  const [isIdle, setIsIdle] = useState<boolean>(true);
  const [isWorking, setIsWorking] = useState<boolean>(false);
  const [isError, setIsError] = useState<boolean>(false);
  const [errorMsg, setErrorMsg] = useState<string | undefined>();
  const [data, setData] = useState<Array<T>>([]);
  const [needRefreshAt, markAsNeedRefreshAt] = useState<number>();

  useEffect(() => {
    (async () => {
      setIsWorking(true);
      setIsError(false);
      setErrorMsg(undefined);
      try {
        const res = await promise();
        setData(res);
      } catch (error) {
        setIsError(true);
        setErrorMsg(error.message);
        setData([]);
      }
      setIsIdle(false);
      setIsWorking(false);
    })();
  }, [needRefreshAt, promise]);

  const refresh = useCallback(() => {
    setIsIdle(true);
    setData([]);
    markAsNeedRefreshAt(new Date().getMilliseconds());
  }, []);

  return {
    isIdle,
    isWorking,
    isError,
    errorMsg,
    data,
    refresh,
    needRefreshAt,
  };
}

export function getGoogleMapUrlByLatLng(lat: string, lng: string) {
  return `${GOOGLE_MAP_URL}${lat},${lng}`;
}
