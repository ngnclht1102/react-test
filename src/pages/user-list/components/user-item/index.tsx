import { TUser } from '@pages/user-list/type';
import { getGoogleMapUrlByLatLng } from '../../../user-list/utils';
import React from 'react';

type TUserItemProps = {
  user: TUser;
};

export const UserItem = ({ user }: TUserItemProps) => {
  const googleMapUrl = getGoogleMapUrlByLatLng(user.address.geo.lat, user.address.geo.lng);

  return (
    <div className="hover:shadow-lg hover:bg-pink-500 rounded-lg bg-pink-400 mt-5 p-5 flex items-center justify-between">
      <div>
        <p data-testid="user-name" className="font-bold text-lg text-white">Name: {user?.name}</p>
        <p data-testid="company-name" className="text-md text-gray-100">Company: {user?.company?.name}</p>
        <p data-testid="address" className="text-md text-gray-100">
          {`Adress: 
          ${user?.address?.suite ?? ''},
          ${user?.address?.street ?? ''}, 
          ${user?.address?.zipcode ?? ''},
          ${user?.address?.city ?? ''}`}
        </p>
      </div>
      <a
        className="text-md text-gray-100 underline italic"
        href={`${googleMapUrl}`}
        target="_blank"
        rel="noopener noreferrer"
      >
        View on map
      </a>
    </div>
  );
};
