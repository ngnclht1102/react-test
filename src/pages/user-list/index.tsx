import React from 'react';
import { useQuery } from '@pages/user-list/utils';
import { UserItem } from './components/user-item';
import { Services } from './services';
import { TUser } from './type';

export default (): JSX.Element => {
  const query = useQuery(Services.getUsers);
  if (query.isWorking || query.isIdle) {
    return <div id="loading">loading</div>;
  }
  if (query.isError) {
    return (
      <div id="error">
        <h1>Error while loading</h1>
        <button
          id="refresh-button"
          className="text-xl font-bold outline outline-offset-2 outline-1"
          onClick={query.refresh}
        >
          Refresh
        </button>
      </div>
    );
  }
  if (!query.data?.length && !query.isWorking && !query.isIdle) {
    return <div>Empty, there are no users.</div>;
  }

  return (
    <div className="container mx-auto w-4/5 my-10" id="container">
      <h1 className="text-xl font-bold">The list of users:</h1>
      <button
        id="refresh-button"
        className="text-xl font-bold outline outline-offset-2 outline-1"
        onClick={query.refresh}
      >
        Refresh
      </button>
      {query.data?.map((user: TUser) => {
        return <UserItem key={user.id} user={user} />;
      })}
    </div>
  );
};
