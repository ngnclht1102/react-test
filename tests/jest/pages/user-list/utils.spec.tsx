import { renderHook, act } from '@testing-library/react-hooks';
import { getGoogleMapUrlByLatLng, useQuery } from '../../../../src/pages/user-list/utils';
import { testUser } from './user-item.spec';
import { BASE_URL, GOOGLE_MAP_URL } from '../../../../src/pages/user-list/constants';
import { fetchMock } from '../../../../utils/fetch-mock';

test('useQuery should return correct initial state', () => {
  const fetcher = () => fetchMock(BASE_URL, [testUser, testUser]);
  const { result } = renderHook(() => useQuery(fetcher));
  expect(result.current.isIdle).toBe(true);
  expect(result.current.isError).toBe(false);
  expect(result.current.isWorking).toBe(true);
  expect(result.current.data).toStrictEqual([]);
  expect(result.current.errorMsg).toBe(undefined);
});

test('useQuery should return correct state in case API call success', async () => {
  const fetcher = () => fetchMock(BASE_URL, [testUser, testUser]);
  const { result, waitFor } = renderHook(() => useQuery(fetcher));
  await waitFor(() => result.current.isWorking === false);
  expect(result.current.isIdle).toBe(false);
  expect(result.current.isWorking).toBe(false);
  expect(result.current.data).toStrictEqual([testUser, testUser]);
  expect(result.current.isError).toBe(false);
  expect(result.current.errorMsg).toBe(undefined);
});

test('useQuery should return correct state in case API call error', async () => {
  const fetcher = () => fetchMock(BASE_URL, [], true);
  const { result, waitFor } = renderHook(() => useQuery(fetcher));

  await waitFor(() => result.current.isWorking === false);
  expect(result.current.isIdle).toBe(false);
  expect(result.current.isWorking).toBe(false);
  expect(result.current.data).toStrictEqual([]);
  expect(result.current.isError).toBe(true);
  expect(result.current.errorMsg).toBeDefined();
});

test('useQuery should reset data, recall the api on refresh', async () => {
  const fetcher = () => fetchMock(BASE_URL, [testUser, testUser]);
  const { result, waitFor } = renderHook(() => useQuery(fetcher));
  await waitFor(() => result.current.isWorking === false);
  expect(result.current.data).toStrictEqual([testUser, testUser]);
  expect(result.current.needRefreshAt).toBeUndefined();
  act(() => {
    result.current.refresh();
  });
  expect(result.current.needRefreshAt).toBeDefined();
  expect(result.current.data).toStrictEqual([]);
  expect(result.current.isWorking).toBe(true);
  await waitFor(() => result.current.isWorking === false);
  expect(result.current.data).toStrictEqual([testUser, testUser]);
});

test('getGoogleMapUrlByLatLng should render correct url', () => {
  const lat = '10,222',
    lng = '102,222';
  const renderedUrl = getGoogleMapUrlByLatLng(lat, lng);
  expect(renderedUrl).toStrictEqual(`${GOOGLE_MAP_URL}${lat},${lng}`);
});
