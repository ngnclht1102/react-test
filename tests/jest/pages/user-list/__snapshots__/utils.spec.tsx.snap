// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User item component 1. Should render component properly 1`] = `
<div
  className="hover:shadow-lg hover:bg-pink-500 rounded-lg bg-pink-400 mt-5 p-5 flex items-center justify-between"
>
  <div>
    <p
      className="font-bold text-lg text-white"
      data-testid="user-name"
    >
      Name: 
      Leanne Graham
    </p>
    <p
      className="text-md text-gray-100"
      data-testid="company-name"
    >
      Company: 
      Romaguera-Crona
    </p>
    <p
      className="text-md text-gray-100"
      data-testid="address"
    >
      Adress: 
          Apt. 556,
          Kulas Light, 
          92998-3874,
          Gwenborough
    </p>
  </div>
  <a
    className="text-md text-gray-100 underline italic"
    href="https://maps.google.com/?q=-37.3159,81.1496"
    rel="noopener noreferrer"
    target="_blank"
  >
    View on map
  </a>
</div>
`;
